services/QuillLMS/app/queries/quill_big_query.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+# All messages to BigQuery should use this module
+
+require 'google/api_client'
+require 'googleauth'
+require 'stringio'
+
+module QuillBigQuery
+  PROJECT_ID = 'analytics-data-stores'
+  FLOAT_FIELDS = %w(correct incorrect percentage)
+
+  class UnsupportedSchemaError < StandardError; end
+
+  def self.valid_schema?(json_body)
+    return false unless json_body['schema']['fields'].respond_to?(:count)
+    return false unless json_body['rows'].respond_to?(:count)
+
+    true
+  end
+
+  def self.floatify_fields(array_of_hashes)
```

Comment:
This transform is specific to `ProgressReports::DistrictConceptReports`. As I add more queries to this framework, I plan to abstract idiosyncratic transforms into procs, which QuillBigQuery will apply as a post-processing step.

Reply:
I think we should reconsider this approach. Asking developers to define a post-query transform proc for every query has a lot of downsides (adds overhead to the query process, it will be hard to keep approaches consistent).

However, the API returns a schema definition for the result. I think it will be more simple and consistent to do basic transforms based on this schema rather than custom procs for each query. BigQuery only has [a handful of types](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types), we can just write transforms for each once, then every query would return results in the correct type. It's more work upfront, but will make this easier to use long-term.